src/common/badblock.c

```diff
@@ -0,0 +1,331 @@
+			fe_end = fe_beg + fmap->fm_extents[e].fe_length - 1;
+
+			if (bb_beg > fe_end || fe_beg > bb_end)
+				continue;
+
+			beg = (bb_beg > fe_beg) ? bb_beg : fe_beg;
+			end = (bb_end < fe_end) ? bb_end : fe_end;
+
+			bbvp = Realloc(bbvp,
+				(++bb_count) * sizeof(struct onebadblock));
+
+			bbvp[bb_count - 1].length = (unsigned)(end - beg + 1);
+			bbvp[bb_count - 1].offset =
+					beg + fmap->fm_extents[e].fe_logical
+					- fmap->fm_extents[e].fe_physical;
+		}
+
+	if (bbs->bbv)
+		Free(bbs->bbv);
+
+	bbs->bbc = bb_count;
+	bbs->bbv = bbvp;
+
+exit_free_all:
+	if (fmap)
+		Free(fmap);
+
+	LOG(10, "bad blocks detected: %u", bbs ? bbs->bbc : 0);
+
+	return bbs;
+}
+
+/*
+ * os_badblocks_clear_file -- clears bad blocks in the regular file
+ *                            (not in a dax device)
+ */
+static int
+os_badblocks_clear_file(const char *file)
+{
+	LOG(3, "file %s", file);
+
+	unsigned long long bb_beg, bb_end;
+	unsigned long long fe_beg, fe_end;
+	off_t beg, end, off, len, not_block_aligned;
+
+	struct badblocks *bbs = NULL;
+	struct fiemap *fmap = NULL;
+	long blksize;
+	int fd;
+
+	if ((fd = os_open(file, O_RDWR)) < 0)
+		return -1;
+
+	if (os_badblocks_extents_common(file, &bbs, &fmap, &blksize)) {
+		close(fd);
+		return -1;
+	}
+
+	if (bbs->bbc == 0 || fmap->fm_extent_count == 0)
+		goto exit_free_all;
+
+	unsigned b, e;
+	for (b = 0; b < bbs->bbc; b++)
+		for (e = 0; e < fmap->fm_extent_count; e++) {
+			bb_beg = bbs->bbv[b].offset;
+			bb_end = bb_beg + bbs->bbv[b].length - 1;
+
+			fe_beg = fmap->fm_extents[e].fe_physical;
+			fe_end = fe_beg + fmap->fm_extents[e].fe_length - 1;
+
+			if (bb_beg > fe_end || fe_beg > bb_end)
+				continue;
+
+			beg = (off_t)((bb_beg > fe_beg) ? bb_beg : fe_beg);
+			end = (off_t)((bb_end < fe_end) ? bb_end : fe_end);
+
+			len = end - beg + 1;
+			off = beg + (off_t)(fmap->fm_extents[e].fe_logical
+					- fmap->fm_extents[e].fe_physical);
+
+			/* check if off is block-aligned */
+			not_block_aligned = off & (blksize - 1);
+			if (not_block_aligned) {
+				beg -= not_block_aligned;
+				off -= not_block_aligned;
+				len += not_block_aligned;
+			}
+
+			/* check if len is block-aligned */
+			not_block_aligned = len & (blksize - 1);
+			if (not_block_aligned) {
+				len += blksize - not_block_aligned;
+			}
+
+			LOG(10,
+				"clearing bad block: physical offset %lu logical offset %li length %li (sectors)",
+				beg >> 9, off >> 9, len >> 9);
+
+			/* deallocate bad blocks */
+			if (fallocate(fd, FALLOC_FL_PUNCH_HOLE |
+					FALLOC_FL_KEEP_SIZE, off, len)) {
+				perror("fallocate");
+			}
+
+			/* allocate new blocks */
+			if (fallocate(fd, FALLOC_FL_KEEP_SIZE, off, len)) {
+				perror("fallocate");
+			}
+		}
+
+exit_free_all:
+	if (fmap)
+		Free(fmap);
+
+	if (bbs)
+		Free(bbs);
+
+	close(fd);
+
+	return 0;
+}
+
+/*
+ * os_badblocks_clear -- clears bad blocks in a file
+ *                      (regular file or dax device)
+ */
+int
+os_badblocks_clear(const char *file)
```

Comment:
If there are more "common" things planned for badblock related things, I'd prefer to split this further:
- devdax badblocks
- file badblocks
- poolset badblocks
and `badblock.h` module that would include all of the above and do this selection and potentially other "common" stuff (for example, the recovery code?).

But if you think this is all these implementations are going to share, I'd move this code to the 'poolset' implementation (which would then include both file and devdax).
Just a suggestion.